cql3/statements/call_statement.cc

```diff
@@ -0,0 +1,141 @@
+                _cf_name->get_column_family(), make_shared<cql3::column_identifier>(col.first, false),
+                col.second);
+
+            columns.push_back(std::move(result_column));
+        }
+        _metadata = seastar::make_shared<cql3::metadata>(std::move(columns));
+    }
+}
+
+uint32_t call_statement::get_bound_terms() const {
+    return _bound_terms;
+}
+
+future<> call_statement::check_access(service::storage_proxy& proxy, const service::client_state& state) const {
+    // todo: check auth::CALL permission
+    return make_ready_future<>();
+}
+
+void
+call_statement::validate(service::storage_proxy&, const service::client_state& state) const {
+    // todo
+}
+
+bool call_statement::depends_on_column_family(const sstring& cf_name) const {
+    return false;
+}
+
+bool call_statement::depends_on_keyspace(const sstring& ks_name) const {
+    // todo : return true if ks_name = procedure keyspace
+    return false;
+}
+
+future<::shared_ptr<cql_transport::messages::result_message>>
+call_statement::execute(query_processor& qp, service::query_state& qs, const query_options& options) const {
+
+    inc_cql_stats(qs.get_client_state().is_internal());
+
+    db::builtin_routine::parameter_map params(_params.begin(), _params.end());
+
+    return _proc->call(std::move(params)).then([this](db::builtin_routine::output output) {
+        using rm = cql_transport::messages::result_message;
+        using srm = ::shared_ptr<rm>;
+        if (_proc->has_output()) {
+            auto result_set = std::make_unique<cql3::result_set>(_metadata);
+            for (auto &&row : output) {
+                std::vector<bytes_opt> rs_row;
+                rs_row.reserve(row.size());
+                for (size_t i = 0; i < row.size(); ++i) {
+                    rs_row.emplace_back(_proc->metadata[i].second->decompose(row[i]));
+                }
+                result_set->add_row(std::move(rs_row));
+            }
+            cql3::result result(std::move(result_set));
+            return make_ready_future<srm>(seastar::make_shared<rm::rows>(std::move(result)));
+        } else {
+            return make_ready_future<srm>(srm{});
+        }
+    });
+}
+
+void call_statement::inc_cql_stats(bool is_internal) const {
+    // todo: proper stats
+}
+
+namespace raw {
+
+std::unique_ptr<prepared_statement>
+call_statement::prepare(database& db, cql_stats& stats) {
+
+    auto& prepared_context = get_prepare_context();
+    seastar::shared_ptr<db::builtin_routine> proc;
+    if (_cf_name->get_keyspace() == "system") {
+        proc = db.get_builtin_routine_registry().find_routine(_cf_name->get_column_family());
+    }
+    if (!proc) {
+        throw exceptions::invalid_request_exception(format("Procedure {} was not found", _cf_name));
+    }
+    auto stmt = ::make_shared<cql3::statements::call_statement>(::make_shared<cf_name>(_cf_name.value()), proc,
+        prepared_context.bound_variables_size(), std::move(_params), stats);
```

Comment:
raw `call_statement` object will be defunct after preparation because of this `move`. The caller looks OK as is, but I'd enjoy some comment warning about this.

Reply:
We should (independently of this series) mark prepare() as consuming its input with `&&` so this is less scary.